### PR TITLE
Support preselect_filters without filter_box IDs

### PR DESCRIPTION
**Pull request by jaylindquist**
We have a use case where external applications build links to dashboards with filters. Currently that requires these applications to know internal database IDs for filter boxes in order to build the `preselect_filters` query parameter.

A second use case we have is users saving links to dashboards with `preselect_filters` already applied with the filter box ID.

> `.../superset/dashboard/dash/?preselect_filters={"414":{"column":["value"]}}`

When moving between environments or during normal maintenance, the filter box IDs may change causing external applications to update and breaking saved links.

This change supports `preselect_filters` without filter box IDs allowing:
* External applications generating filters for dashboards without needing database IDs
* "Share Dashboard" button generating share links without dashboard filters, letting them continue to work if filter box IDs change
* Backwards compatibility of `preselect_filters` allowing existing saved links to function

> `.../superset/dashboard/dash/?preselect_filters={"column":["value"]}`

When a dashboard is loaded with the above `preselect_filters` the `getInitialState()` function for dashboards will search through all filter boxes that support `column` as a filter and set it to `["value"]` as its value. Additionally, `preselect_filters` is backwards compatible, supporting the original format as well as the new format.

### CODE REVIEW

**Comment by graceguo-supercat** [comment]
hi @jaylindquist, i like the idea of simplifying pre-selected filters. But what if dashboard has two filter box, and both of them have `column` key, but you only want to apply 1 filter box?

**Comment by jaylindquist** [comment]
@graceguo-supercat All existing functionality is still supported for `preselect_filters`. However, the 'Share Dashboard' generated link was changed to no longer use filter box IDs.

For the current implementation of Superset, if you have two filter boxes, both with a 'country' filter, selecting 'Bhutan' in one filter box still applies the filter across all charts -- including other filter boxes (the other filter box is blank, but only allows you to select 'Bhutan' as a filter). This all works as before. 

When visiting the dashboard with`preselect_filters={"75":{"country":["Bhutan"]}}`, current functionality will set filter box 75 'country' to 'Bhutan' while leaving the other filter box blank, but only able to select 'Bhutan'. This functionality is also still supported in the PR when `preselect_filters` uses the format with IDs.

Existing functionality in version 0.28.1:
v0 28 1-filter

The noticeable change is when visiting the dashboard with `preselect_filters={"country":["Bhutan"]}`. In that case, both filter boxes will have their 'country' filter set to 'Bhutan'.

New functionality in this PR:
pr-filter

**Comment by jaylindquist** [comment]
@graceguo-supercat I manually tested the filter_immune_slices and it continues to work as expected.

Since the changes I made only really affect the URL and not any internal representation of the filters (`getInitialState` just translates the new URL format to the original internal filter format, while `getDashboardUrl` translates the internal filter format to the new URL format) I hadn't made any changes to how Superset is storing this information. `default_filters` works as it always has and can stay the same going forward.

Do you want `default_filters` for existing and new dashboards updated? If so, a DB migration script is needed as well as changes to how dashboards are saved.

File: superset/assets/spec/javascripts/dashboard/reducers/getInitialState_spec.js [hunk 1]
```diff
@@ -0,0 +1,166 @@
+import getInitialState from '../../../../src/dashboard/reducers/getInitialState';
```
Comment (bolkedebruin) [inline superset/assets/spec/javascripts/dashboard/reducers/getInitialState_spec.js]: Please add ASF license header

File: superset/assets/src/dashboard/util/filterOptions.js [hunk 2]
```diff
@@ -0,0 +1,10 @@
+// maps control names to their key in extra_filters
```
Comment (bolkedebruin) [inline superset/assets/src/dashboard/util/filterOptions.js]: Idem, please add ASF header

File: superset/assets/src/dashboard/util/getDashboardUrl.js [hunk 3]
```diff
@@ -1,6 +1,32 @@
 /* eslint camelcase: 0 */
```
Comment (bolkedebruin) [inline superset/assets/src/dashboard/util/getDashboardUrl.js]: Add header ASF please

**Review by bolkedebruin — CHANGES_REQUESTED** [review]

File: superset/assets/src/dashboard/util/filterOptions.js (hunk 2 above)
Comment (jaylindquist) [reply superset/assets/src/dashboard/util/filterOptions.js]: done

File: superset/assets/src/dashboard/util/getDashboardUrl.js (hunk 3 above)
Comment (jaylindquist) [reply superset/assets/src/dashboard/util/getDashboardUrl.js]: done

**Review by bolkedebruin — APPROVED** [review]

**Comment by graceguo-supercat** [comment]
@jaylindquist I am thinking db migration is a necessary, as well as updating dashboard save `default_filters` metadata. Otherwise user will `always` see 2 patterns in preselect_filters parameter, and will feel confused. 
At the same time, even with db migration that converts old formation into new format, in the frontend code we still have to handle both patterns. I feel some users use API to generate dashboard requests, or may copy urls somewhere.

Thanks for bringing the feature and code cleanup.

**Comment by jaylindquist** [comment]
@graceguo-supercat I've added a database migration and incorporated the single-value filters that were added

**Pull request closed without merging**